src/main/java/net/objecthunter/exp4j/tokenizer/Tokenizer.java

```diff
@@ -133,6 +148,16 @@ private boolean isCloseParentheses(char ch) {
         return ch == ')' || ch == '}' || ch == ']';
     }
 
+    private Collection<String> undefinedVariables = null;
+    
+    public void allowUndefinedVariables(boolean allow) {
+    	this.undefinedVariables = (allow ? new ArrayList<String>() : null);
+    }
+    
+    public String[] getUndefinedVariables() {
+    	return (undefinedVariables != null ? undefinedVariables.toArray(new String[undefinedVariables.size()]) : null);
```

Comment:
The conclusion after reading [_Arrays of Wisdom of the Ancients_](http://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion) is that we should use ```toArray(new T[0])``` instead of ```toArray(new T[size])```:

> Bottom line: toArray(new T[0]) seems faster, safer, and contractually cleaner, and therefore should be the default choice now. Future VM optimizations may close this performance gap for toArray(new T[size]), rendering the current "believed to be optimal" usages on par with an actually optimal one.